### PR TITLE
[GraphBolt] Implement proper `IndexSelectCSC` for CPU.

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Before, the CPU IndexSelectCSC was going through InSubgraph of CPU. Now, it is the other way around as it should be. I am trying to clean up the messes from before.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/src/index_select.cc [hunk 1]
```diff
@@ -129,17 +132,78 @@ std::tuple<torch::Tensor, std::vector<torch::Tensor>> IndexSelectCSCBatched(
...
+              for (const auto& indices : indices_list) {
+                results.push_back(torch::empty(
+                    *output_size,
+                    nodes.options().dtype(indices.scalar_type())));
+              }
+              if (with_edge_ids) {
+                edge_ids = torch::empty(
+                    *output_size, nodes.options().dtype(indptr.scalar_type()));
+              }
+              torch::parallel_for(
```
Comment (Rhett-Ying) [inline graphbolt/src/index_select.cc:176]: This change obtains 20x improvement in regression. This mainly benefit from the change from `torch.cat()` to `pre-allocate for output tensor + copy indices from source directly to target positions of output tensor`, right? Namely, benefit from avoiding large number of memory allocation and de-allocation? @mfbalin

Comment (mfbalin) [reply graphbolt/src/index_select.cc:176]: That and overall not doing any wasted work.

File: graphbolt/src/fused_csc_sampling_graph.cc [hunk 2]
```diff
@@ -293,48 +294,21 @@ c10::intrusive_ptr<FusedSampledSubgraph> FusedCSCSamplingGraph::InSubgraph(
...
+  }
 
   return c10::make_intrusive<FusedSampledSubgraph>(
-      indptr.cumsum(0), torch::cat(indices_arr), torch::cat(edge_ids_arr),
-      nodes, torch::arange(0, NumNodes()),
-      type_per_edge_
-          ? torch::optional<torch::Tensor>{torch::cat(type_per_edge_arr)}
-          : torch::nullopt);
+      output_indptr, results.at(0), results.back(), nodes,
+      torch::arange(0, NumNodes()), type_per_edge);
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply graphbolt/src/fused_csc_sampling_graph.cc:311]: supposed to be original_row_id: https://github.com/dmlc/dgl/blob/ff088ac9991ec81e432a8629d9553cdead59aa8e/graphbolt/include/graphbolt/fused_sampled_subgraph.h#L50
But seems not used in `g.in_subgraph()` in python.

[the remark this answers is not in the document]
Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:311]: We probably want to make it nullopt then. SampleNeighbors does the same I think.

[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply graphbolt/src/fused_csc_sampling_graph.cc:311]: I think so as we don't really use it for now.

[the remark this answers is not in the document]
Comment (mfbalin) [reply graphbolt/src/fused_csc_sampling_graph.cc:311]: Then the regression test might even see more of a speedup after cleaning this up.

**Pull request merged**